File: rules/emptypassword.js
const fs = require('fs');
const vscode = require('vscode');


var smell = {

    detect : (fileName, token) => {
        try{
            if(token.hasOwnProperty("line")) var lineno = token.line;
            if(token.hasOwnProperty("type")) var tokenType = token.type;
            
            const MSG = 'possible use of empty password'
            
            const WARNING_MSG = MSG+' at line '+ lineno;
            const commonPasswords = ['password','pwd','userpassword','userpwd', 'userpass', 'pass_no', 'pass-no','user-pass', 'upass', 'user_pass', 'u_pass',  
                                    'usr_pwd','usr_pass', 'usr-pass','userpasswords', 'user-passwords', 'user-password', 'user_password', 'use_pass','user_pwd'
                                ]
    
            if(token.type == "variable" && token.name != null && (token.value == null || token.value.length == 0) && token.valueSrc == 'initialization'){
                for(const pwd of commonPasswords){
                    let prefixMatch = new RegExp(`\\b${pwd}[_A-Za-z0-9-]*`);
                    let suffixMatch = new RegExp(`[_A-Za-z0-9-]*${pwd}\\b`);
    
                    if(token.name.toLowerCase().match(prefixMatch)) {
                        smell.triggerAlarm (fileName, MSG, lineno, WARNING_MSG);
                        break
                    }
                    else if(token.name.toLowerCase().match(suffixMatch)){ 
                        smell.triggerAlarm (fileName, MSG, lineno, WARNING_MSG);
                        break
                    }
                }
            }    
            
            else if(token.type == 'dict' && token.hasOwnProperty('pairs')){
                
                token.pairs.forEach( pair => {
                    for (const pwd of commonPasswords){
    
                        let suffixMatch = new RegExp(`[_A-Za-z0-9-]*${pwd}\\b`);
                        
                        if(pair[0].toLowerCase().match(suffixMatch) && (pair[1] == null || pair[1].length == 0)){
                            smell.triggerAlarm (fileName, MSG, lineno, WARNING_MSG);
                            break
                        } 
                    }
                })
            } 
    
            else if(tokenType == 'comparison' && token.hasOwnProperty('pairs')){
                token.pairs.forEach( pair => {
                    for (const pwd of commonPasswords){
            
                        let re = new RegExp(`[_A-Za-z0-9-]*${pwd}\\b`);
                        if(pair[0].toLowerCase().match(re) && (pair[1] == null || pair[1].length == 0)) smell.triggerAlarm (fileName, MSG, lineno, WARNING_MSG);
                    }
                })
            }
            else if(tokenType == 'function_def' && token.hasOwnProperty('args') && token.hasOwnProperty('defaults')){
                let argsLength = token.args.length
                let defaultsLength = token.defaults.length;
                
                let args = token.args.slice(argsLength - defaultsLength, argsLength)
                let defaults = token.defaults
    
                for(let i = 0; i< args.length; i++){
                    for (const pwd of commonPasswords){
            
                        let re = new RegExp(`[_A-Za-z0-9-]*${pwd}\\b`);
                        if(args[i].toLowerCase().match(re) && (defaults[i][0] == null || defaults[i][0].length == 0) && defaults[i][1] == true) 
                            smell.triggerAlarm (fileName, MSG, lineno, WARNING_MSG);
                    }
                }  
            }
            else if(token.type == 'function_call' && token.hasOwnProperty('keywords')){
                
                token.keywords.forEach( keyword => {
                    for(const pwd of commonPasswords){
    
                        let re = new RegExp(`[_A-Za-z0-9-]*${pwd}\\b`)
                        if(token.keywords.length == 3 && keyword[0].match(re) && (keyword[1][0] == null || keyword[1][0].length == 0) && keyword[1][1] == true){
                            smell.triggerAlarm (fileName, MSG, lineno, WARNING_MSG);
                            break
                        }   
                    }
                })
            }
        } catch(error){
            console.log(error);
        }
    },
    
    triggerAlarm: (fileName, MSG, lineno, WARNING_MSG) => {
        let backslashSplittedFilePathLength = fileName.split("/").length
        let filenameFromPath = fileName.split("/")[backslashSplittedFilePathLength - 1]
        
        vscode.window.showWarningMessage(MSG +" : "+ filenameFromPath+":"+lineno);
        console.log( "\u001b[1;31m"+"warning: "+MSG +"  location:"+ fileName+":"+lineno);
        fs.appendFileSync(__dirname+'/../warning-logs/project_warnings.csv', fileName+" ,"+WARNING_MSG+"\n");
        
        // fs.appendFile(__dirname+'/../logs/project_warnings.csv', fileName+","+WARNING_MSG+"\n", (err) => err ? console.log(err): "");
    }
}

module.exports = smell;

File: py-scripts/filters/emptypassword.py
import re

class EmptyPassword:
    '''This is the class for detecting empty passwords in code'''

    def __init__(self):
        self.common_passwords = ['password','pass','pwd','passwd', 'upass']
        self.warning_message = 'empty password'

    def detect_smell(self, token, src_file):
        try:
            if token.__contains__("line"): lineno = token["line"]
            if token.__contains__("type"): token_type = token["type"]
            if token.__contains__("name"): name = token["name"]
            if token.__contains__("value"): value = token["value"]
            if token.__contains__("valueSrc"): valueSrc = token["valueSrc"]

            
            if( 
                token_type == "variable" 
                and name is not None
                and isinstance(value, str)  
                and (value is None or len(value) == 0) 
                and token.__contains__('values') is False
                and valueSrc == "initialization"
            ):  
                for pwd in self.common_passwords:
                    if (re.match(r'[_A-Za-z0-9-\.]*{pwd}\b'.format(pwd = pwd), name.lower().strip()) or re.match(r'\b{pwd}[_A-Za-z0-9-\.]*'.format(pwd = pwd), name.lower().strip())):
                        self.trigger_alarm(src_file, lineno)
                        break


            elif token_type == "dict" and name is not None and token.__contains__("pairs"):
                
                for pair in token['pairs']:
                    for pwd in self.common_passwords:
                        if(
                            isinstance(pair[0], str) 
                            and (re.match(r'[_A-Za-z0-9-\.]*{pwd}\b'.format(pwd = pwd), pair[0].lower().strip()) or re.match(r'\b{pwd}[_A-Za-z0-9-\.]*'.format(pwd = pwd), pair[0].lower().strip())) 
                            and (pair[1] is None or len(pair[1]) == 0)
                        ):   
                            self.trigger_alarm(src_file, lineno)
            

            elif token_type == "comparison" and token.__contains__("pairs"):
                
                for pair in token["pairs"]:
                    for pwd in self.common_passwords:
                        if(
                            isinstance(pair[0], str)
                            and (re.match(r'[_A-Za-z0-9-\.]*{pwd}\b'.format(pwd = pwd), pair[0].lower().strip()) or re.match(r'\b{pwd}[_A-Za-z0-9-\.]*'.format(pwd = pwd), pair[0].lower().strip())) 
                            and (pair[1] is None or (isinstance(pair[1], str) and len(pair[1]) == 0))
                        ):    
                            self.trigger_alarm(src_file, lineno)
                            break

            elif token_type == "function_call" and token.__contains__('keywords'):
                for keyword in token['keywords']:
                    for pwd in self.common_passwords:
                
                        if (
                            len(keyword) == 3
                            and isinstance(keyword[0], str) 
                            and re.match(r'[_A-Za-z0-9-]*{pwd}\b'.format(pwd = pwd), keyword[0].lower().strip()) 
                            and (keyword[1] is None or len(str(keyword[1])) == 0) and keyword[2] is True
                        ): 
                                self.trigger_alarm(src_file, lineno)
                                break

            elif token_type == "function_def" and token.__contains__("args") and token.__contains__("defaults"):
                defaults_size = len(token['defaults'])
                args = token['args'][-defaults_size:]

                for pair in zip(args, token['defaults']):
                        for pwd in self.common_passwords:
                            if re.match(r'[_A-Za-z0-9-]*{pwd}\b'.format(pwd = pwd), pair[0].lower().strip()) and (pair[1][0] == None or len(str(pair[1][0])) == 0) and pair[1][1] is True: 
                                self.trigger_alarm(src_file, lineno)
                                break

        except Exception as error:
            print(str(error))

    def trigger_alarm(self, src_file, lineno):
        print(src_file +":"+ str(lineno)+" ,"+self.warning_message)
    